src/adm.rs

```diff
@@ -1,43 +1,275 @@
+            return Err(HandlerErrorKind::InvalidHost(species, url.to_string()).into());
+        }
+    };
+    let host = match parsed.host() {
+        Some(v) => v.to_string(),
+        None => {
+            tags.add_tag("type", species);
+            tags.add_extra("url", &url);
+            return Err(HandlerErrorKind::MissingHost(species, parsed.to_string()).into());
+        }
+    };
+    if !filter.contains(&host) {
+        tags.add_tag("type", species);
+        tags.add_extra("url", &url);
+        return Err(HandlerErrorKind::UnexpectedHost(species, host).into());
+    }
+    Ok(())
+}
+
+impl AdmFilter {
+    /// Report the error directly to sentry
+    fn report(&self, error: &HandlerError, tags: &Tags) {
+        // dbg!(&error, &tags);
+        // TODO: if not error.is_reportable, just add to metrics.
+        l_sentry::report(tags, sentry::event_from_error(error));
+    }
+
+    /// Check the advertiser URL
+    fn check_advertiser(
+        &self,
+        filter: &AdmAdvertiserFilterSettings,
+        tile: &mut AdmTile,
+        tags: &mut Tags,
+    ) -> HandlerResult<()> {
+        check_url(
+            &tile.advertiser_url,
+            "Advertiser",
+            &filter.advertiser_hosts,
+            tags,
+        )
+    }
+
+    /// Check the click URL
+    fn check_click(
+        &self,
+        filter: &AdmAdvertiserFilterSettings,
+        tile: &mut AdmTile,
+        tags: &mut Tags,
+    ) -> HandlerResult<()> {
+        check_url(&tile.click_url, "Click", &filter.click_hosts, tags)
+    }
+
+    /// Check the impression URL to see if it's valid.
+    ///
+    /// This extends `filter_and_process`
+    fn check_impression(
+        &self,
+        filter: &AdmAdvertiserFilterSettings,
+        tile: &mut AdmTile,
+        tags: &mut Tags,
+    ) -> HandlerResult<()> {
+        check_url(
+            &tile.impression_url,
+            "Impression",
+            &filter.impression_hosts,
+            tags,
+        )
+    }
+
+    /// Filter and process tiles from ADM:
+    ///
+    /// - Returns None for tiles that shouldn't be shown to the client
+    /// - Modifies tiles for output to the client (adding additional fields, etc.)
+    pub fn filter_and_process(&self, tile: AdmTile, tags: &mut Tags) -> Option<AdmTile> {
+        // Use strict matching for now, eventually, we may want to use backwards expanding domain
+        // searches, (.e.g "xyz.example.com" would match "example.com")
+        let mut result = tile.clone();
+        match self.filter_set.get(&tile.name.to_lowercase()) {
+            Some(filter) => {
+                // Apply any additional tile filtering here.
+                let none = AdmAdvertiserFilterSettings::default();
+                let default = self
+                    .filter_set
+                    .get(&DEFAULT.to_lowercase())
+                    .unwrap_or(&none);
+                let adv_filter = if filter.advertiser_hosts.is_empty() {
```

Comment:
looks like these could all use the `Option::or` like `result.position` below

Reply:
I don't think `Vec` has an `.or()`. I could do a `filter.advertiser_hosts.append(default.clone())`, but that alters the behavior to always include the `default` instead of "use default if they haven't defined anything".

Reply:
My bad, I read these as Options